src/main/java/org/prebid/server/execution/RemoteFileSyncer.java

```diff
@@ -150,67 +201,84 @@ private long setTimeoutTimer(AsyncFile asyncFile, Future<Void> future, Pump pump
             cleanUp().compose(aVoid -> download())
-                    .setHandler(retryResult -> handleRetryResult(receivedFuture, retryInterval, next, retryResult));
+                    .setHandler(retryResult -> handleRetryResult(retryResult, retryInterval, next, receivedFuture));
         } else {
-            cleanUp().setHandler(aVoid -> receivedFuture.fail(new PreBidException("File sync failed after retries")));
+            cleanUp().setHandler(aVoid -> receivedFuture.fail("File sync failed after retries"));
         }
     }
 
     private Future<Void> cleanUp() {
-        final Future<Void> future = Future.future();
-        fileSystem.exists(saveFilePath, existResult -> handleFileExistsWithDelete(future, existResult));
-        return future;
+        return checkFileExist().compose(fileExists -> fileExists ? deleteTmpFile() : Future.succeededFuture());
     }
 
-    private void handleFileExistsWithDelete(Future<Void> future, AsyncResult<Boolean> existResult) {
-        if (existResult.succeeded()) {
-            if (existResult.result()) {
-                fileSystem.delete(saveFilePath, future);
-            } else {
-                future.complete();
-            }
-        } else {
-            future.fail(new PreBidException(String.format("Cant check if file exists %s", saveFilePath)));
-        }
+    private Future<Void> deleteTmpFile() {
+        final Future<Void> result = Future.future();
+        fileSystem.delete(tmpFilePath, result);
+        return result;
     }
 
-    private void handleRetryResult(Future<Void> future, long retryInterval, long next, AsyncResult<Void> retryResult) {
+    private void handleRetryResult(AsyncResult<Void> retryResult, long retryInterval, long next, Future<Void> result) {
         if (retryResult.succeeded()) {
-            future.complete();
+            result.complete();
         } else {
-            retryDownload(future, retryInterval, next);
+            retryDownload(retryInterval, next, result);
         }
     }
 
-    private void handleSync(Consumer<String> consumer, AsyncResult<Void> aVoid) {
-        if (aVoid.succeeded()) {
-            consumer.accept(saveFilePath);
+    private Future<Boolean> tryUpdate() {
+        return checkFileExist()
+                .compose(fileExists -> fileExists ? isNeedToUpdate() : Future.succeededFuture(true))
+                .compose(needUpdate -> needUpdate ? syncFiles().map(true) : Future.succeededFuture(false));
+    }
+
+    private Future<Boolean> isNeedToUpdate() {
+        final Future<Boolean> isNeedToUpdate = Future.future();
+        httpClient.headAbs(downloadUrl, response -> checkNewVersion(response, isNeedToUpdate))
+                .exceptionHandler(isNeedToUpdate::fail)
+                .end();
+        return isNeedToUpdate;
+    }
+
+    private void checkNewVersion(HttpClientResponse response, Future<Boolean> isNeedToUpdate) {
+        final String contentLengthParameter = response.getHeader(HttpHeaders.CONTENT_LENGTH);
```

Comment:
Is content-length a reliable property for assuming that the content hasn't been updated on the server?

Reply:
This is the best solution that I came up with.
It was tested manually